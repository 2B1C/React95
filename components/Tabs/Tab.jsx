import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const NavItem = styled.li``;

class Tab extends Component {
  constructor(props) {
    super(props);
  }

  onClick = () => {
    const { onClick } = this.props;
    onClick(this);
  };

  render() {
    const { activeTab, title } = this.props;

    return (
      <NavItem activeTab={activeTab} onClick={this.onClick}>
        {title}
      </NavItem>
    );
  }
}

Tab.propTypes = {
  activeTab: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default Tab;
