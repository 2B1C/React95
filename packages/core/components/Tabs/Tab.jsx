import React from "react";
import PropTypes from "prop-types";
import styled from "@xstyled/styled-components";
import { th } from "@xstyled/system";

const NavItem = styled.li`
  list-style: none;
  padding: 3 6;
  background-color: grays.2;
  box-shadow: inset 0px 1px 0px 0px ${th("colors.white")},
    -2px 1px 0 -1px ${th("colors.grays.2")},
    -3px 1px 0px -1px ${th("colors.white")},
    -2px 0px 0px -1px ${th("colors.white")},
    2px 1px 0 -1px ${th("colors.grays.3")}, 2px 0px 0 -1px ${th("colors.black")},
    3px 1px 0 -1px ${th("colors.black")};
  margin-right: 2;
  margin-left: 2;

  &:first-child {
    margin-left: 4;
  }

  ${props =>
    props.active &&
    `
      margin-top: -2px;
      margin-left: -2px;
      margin-bottom: 1px;

      &:first-child {
        margin-left: 2px;
      }

      box-shadow: inset 0px 1px 0px 0px #ffffff, -2px 1px 0 -1px #c3c7cb,
        -0.5px 3px 0 0.5px #c3c7cb, -2px 2px 0px 0px #fff,
        -2px 0px 0px -1px #ffffff, 1px 2px 0 0px #868a8e, 2px 0px 0 -1px #000,
        2px 2px 0 0px #000;

      & + li {
        margin-left: 0;
        box-shadow: inset 0px 1px 0px 0px #ffffff, 2px 1px 0 -1px #868a8e,
          2px 0px 0 -1px #000, 3px 1px 0 -1px #000;
      }
    `}
`;

const Tab = ({ activeTab, title, ...rest }) => (
  <NavItem {...rest} active={activeTab === title}>
    {title}
  </NavItem>
);

Tab.propTypes = {
  activeTab: PropTypes.string,
  title: PropTypes.string.isRequired
};

Tab.defaultProps = {
  activeTab: undefined
};

export default Tab;
