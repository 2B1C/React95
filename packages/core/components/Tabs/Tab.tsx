import * as React from 'react';
import styled, { css } from '@xstyled/styled-components';
import { th } from '@xstyled/system';

const NavItem = styled.li<{ active: Boolean }>`
  list-style: none;
  padding: 3 6;
  background-color: bg;
  box-shadow: inset 0 1px 0 0 ${th('colors.white')},
    -2px 1px 0 -1px ${th('colors.grays.2')},
    -3px 1px 0 -1px ${th('colors.white')}, -2px 0 0 -1px ${th('colors.white')},
    2px 1px 0 -1px ${th('colors.grays.3')}, 2px 0 0 -1px ${th('colors.black')},
    3px 1px 0 -1px ${th('colors.black')};
  margin-right: 2;
  margin-left: 2;

  &:first-child {
    margin-left: 4;
  }

  ${({ active }) =>
    active &&
    css`
      margin-top: -${th('space.2')}px;
      margin-left: -${th('space.2')}px;
      margin-bottom: 1;

      &:first-child {
        margin-left: 2;
      }

      box-shadow: inset 0 1px 0 0 ${th('colors.white')},
        -2px 1px 0 -1px ${th('colors.grays.2')},
        -0.5px 3px 0 0.5px ${th('colors.grays.2')},
        -2px 2px 0 0 ${th('colors.white')}, -2px 0 0 -1px ${th('colors.white')},
        1px 2px 0 0 ${th('colors.grays.3')}, 2px 0 0 -1px ${th('colors.black')},
        2px 2px 0 0 ${th('colors.black')};

      & + li {
        margin-left: 0;
        box-shadow: inset 0 1px 0 0 ${th('colors.white')},
          2px 1px 0 -1px ${th('colors.grays.3')},
          2px 0 0 -1px ${th('colors.black')},
          3px 1px 0 -1px ${th('colors.black')};
      }
    `}
`;

export interface TabProps extends React.HTMLAttributes<HTMLLIElement> {
  activeTab?: string;
  title: string;
  disabled?: Boolean;
  onClick?(e: React.MouseEvent): void;
}

const Tab: React.FC<TabProps> = ({ activeTab, title, ...rest }) => (
  <NavItem {...rest} active={activeTab === title}>
    {title}
  </NavItem>
);

export default Tab;
