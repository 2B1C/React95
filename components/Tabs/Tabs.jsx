import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Tab from './Tab';

const Navbar = styled.nav``;

const renderIf = (conditional, renderFn) => conditional && renderFn();

class Tabs extends Component {
  constructor(props) {
    super(props);

    const { children } = props;

    const firstTab = React.Children.toArray(children)[0];

    this.state = {
      activeTab: firstTab.props.title,
    };
  }

  onTabClick = ({ props: { title } }) => {
    this.setState({ activeTab: title });
  };

  render() {
    const { children } = this.props;
    const { activeTab } = this.state;

    return (
      <React.Fragment>
        <Navbar>
          {activeTab}
          <ol>
            {React.Children.map(children, child => {
              const { title } = child.props;

              return (
                <Tab
                  key={title}
                  title={title}
                  activeTab={activeTab}
                  onClick={this.onTabClick}
                />
              );
            })}
          </ol>
        </Navbar>

        {React.Children.map(children, child =>
          renderIf(child.props.title === activeTab, () => child),
        )}
      </React.Fragment>
    );
  }
}

Tabs.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
};

export default Tabs;
