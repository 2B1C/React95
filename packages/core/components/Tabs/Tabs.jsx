import React, { useState } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import Tab from "./Tab";

const Navbar = styled.ol`
  display: flex;
  padding: 0;
  margin: 0;
`;

const NavContainer = styled.div`
  padding: 12px;
  margin-left: 1px;
  background-color: #c3c7cb;
  box-shadow: inset 0px 1px 0px 0px #ffffff, 2px 1px 0 -1px #868a8e,
    1px 1px 0 0px #868a8e, 0px 1px 0 0px #868a8e, 2px 0px 0 -1px #000,
    2px 2px 0 0px #000, 0px 2px 0 0px #000, 0px 1px 0px 1px #ffffff,
    2px 0px 0 0px #000;
`;

NavContainer.displayName = "NavContainer";

const If = ({ condition, children }) => condition && children;

const Tabs = ({ children, style, defaultActiveTab, onChange, ...rest }) => {
  const firstTab = React.Children.toArray(children)[0];
  const [activeTab, setActiveTab] = useState(
    defaultActiveTab || firstTab.props.title
  );

  return (
    <>
      <Navbar style={style} {...rest}>
        {React.Children.map(children, child => {
          const { title, disabled } = child.props;

          return (
            <Tab
              key={title}
              title={title}
              activeTab={activeTab}
              onClick={
                !disabled &&
                (() => {
                  if (onChange) {
                    onChange(title);
                  }
                  setActiveTab(title);
                })
              }
            />
          );
        })}
      </Navbar>

      <NavContainer style={style}>
        {React.Children.map(children, child => (
          <If condition={child.props.title === activeTab}>
            {child.props.children}
          </If>
        ))}
      </NavContainer>
    </>
  );
};

Tabs.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  // eslint-disable-next-line react/forbid-prop-types
  style: PropTypes.object,
  defaultActiveTab: PropTypes.string,
  onChange: PropTypes.func
};

Tabs.defaultProps = {
  children: [],
  style: {},
  defaultActiveTab: undefined,
  onChange: undefined
};

export default Tabs;
