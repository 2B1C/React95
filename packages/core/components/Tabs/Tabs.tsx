import * as React from 'react';
import styled from '@xstyled/styled-components';

import Tab, { TabProps } from './Tab';
import Frame from '../Frame';

const Navbar = styled(Frame)`
  display: flex;
  padding: 0;
  margin: 0;
  border: none;
  box-shadow: none;
  background-color: unset;
`;

const NavContainer = styled(Frame)`
  padding: 12;
  box-shadow: out;
`;

NavContainer.displayName = 'NavContainer';

const If = ({ condition, children }: { condition: Boolean; children: any }) =>
  condition && children;

interface TabsProps {
  style: React.CSSProperties;
  defaultActiveTab?: string;

  onChange?(title: string, e: React.MouseEvent): void;
}

const Tabs: React.FC<TabsProps> = ({
  children,
  style,
  defaultActiveTab,
  onChange,
  ...rest
}) => {
  const [firstTab] = React.Children.toArray(children) as Array<
    React.ReactElement<TabProps>
  >;
  const [activeTab, setActiveTab] = React.useState(
    defaultActiveTab || firstTab.props.title,
  );

  return (
    <>
      <Navbar style={style} {...rest} as="ol">
        {React.Children.map(children, (child: React.ReactElement<TabProps>) => {
          const { title, disabled } = child.props;

          return (
            <Tab
              key={title}
              {...child.props}
              activeTab={activeTab}
              onClick={e => {
                if (!disabled) {
                  if (onChange) {
                    onChange(title, e);
                  }
                  setActiveTab(title);
                }
              }}
            />
          );
        })}
      </Navbar>

      <NavContainer style={style}>
        {React.Children.map(children, (child: React.ReactElement<TabProps>) => (
          <If condition={child.props.title === activeTab}>
            {child.props.children}
          </If>
        ))}
      </NavContainer>
    </>
  );
};

export default Tabs;
