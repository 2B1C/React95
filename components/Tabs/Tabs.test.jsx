import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Tab from './Tabs';
import Tabs from './Tabs';

describe('<Tabs />', () => {
  describe('Snapshot', () => {
    it('should match snapshot', () => {
      expect(
        renderer
          .create(
            <Tabs>
              <Tab title="example">
                <p>Example text</p>
              </Tab>
            </Tabs>
          )
          .toJSON()
      ).toMatchSnapshot();
    });
  });

  describe('Active Tab', () => {
    const firstText = 'First text';
    const secondText = 'Second text';
    const MountedTab = mount(
      <Tabs>
        <Tab title="first">
          <p>{firstText}</p>
        </Tab>
        <Tab title="Second">
          <p>{secondText}</p>
        </Tab>
      </Tabs>
    );

    it('should have first Tab active', () => {
      expect(MountedTab.find('NavContainer').text()).toBe(firstText);
    });

    it('should active second Tab when Tab is clicked', () => {
      expect(MountedTab.find('NavContainer').text()).toBe(firstText);

      MountedTab.find('Tab')
        .at(1)
        .simulate('click');

      expect(MountedTab.find('NavContainer').text()).toBe(secondText);
    });
  });
});
