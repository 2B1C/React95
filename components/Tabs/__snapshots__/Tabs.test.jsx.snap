// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> Snapshot should match snapshot 1`] = `
Array [
  .c1 {
  list-style: none;
  padding: 3px 6px;
  background-color: #c3c7cb;
  box-shadow: inset 0px 1px 0px 0px #ffffff,-2px 1px 0 -1px #c3c7cb,-3px 1px 0px -1px #ffffff,-2px 0px 0px -1px #ffffff,2px 1px 0 -1px #868a8e,2px 0px 0 -1px #000,3px 1px 0 -1px #000;
  margin-right: 2px;
  margin-left: 2px;
  margin-top: -2px;
  margin-left: -2px;
  margin-bottom: 1px;
  box-shadow: inset 0px 1px 0px 0px #ffffff,-2px 1px 0 -1px #c3c7cb,-0.5px 3px 0 0.5px #c3c7cb,-2px 2px 0px 0px #fff,-2px 0px 0px -1px #ffffff,1px 2px 0 0px #868a8e,2px 0px 0 -1px #000,2px 2px 0 0px #000;
}

.c1:first-child {
  margin-left: 4px;
}

.c1:first-child {
  margin-left: 2px;
}

.c1 + li {
  margin-left: 0;
  box-shadow: inset 0px 1px 0px 0px #ffffff,2px 1px 0 -1px #868a8e,2px 0px 0 -1px #000,3px 1px 0 -1px #000;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  margin: 0;
}

<ol
    className="c0"
  >
    <li
      className="c1"
      onClick={[Function]}
    >
      example
    </li>
  </ol>,
  .c0 {
  padding: 12px;
  margin-left: 1px;
  background-color: #c3c7cb;
  box-shadow: inset 0px 1px 0px 0px #ffffff,2px 1px 0 -1px #868a8e,1px 1px 0 0px #868a8e,0px 1px 0 0px #868a8e,2px 0px 0 -1px #000,2px 2px 0 0px #000,0px 2px 0 0px #000,0px 1px 0px 1px #ffffff,2px 0px 0 0px #000;
}

<div
    className="c0"
  >
    <p>
      Example text
    </p>
  </div>,
]
`;
